File: internal/scanner/scanner.go
package scanner

import (
	"bufio"
	"favmaster/internal/downloader"
	"favmaster/internal/hasher"
	"fmt"
	"log"
	"net/url"
	"os"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
)

var (
	SupportedImageExtensions = map[string]bool{
		".ico": true, ".png": true, ".jpg": true, ".jpeg": true,
		".gif": true, ".svg": true, ".webp": true,
	}

	FaviconPaths = []string{
		"/favicon.ico",
		"/favicon.png",
		"/favicon.jpg",
		"/favicon.jpeg",
		"/favicon.svg",
		"/favicon.webp",
		"/images/favicon.ico",
		"/img/favicon.ico",
		"/assets/favicon.ico",
		"/static/favicon.ico",
		"/icon.png",
		"/logo.png",
		"/apple-touch-icon.png",
	}
)

func IsImageURL(url string) bool {
	lowerURL := strings.ToLower(url)
	for ext := range SupportedImageExtensions {
		if strings.HasSuffix(lowerURL, ext) {
			return true
		}
	}
	return false
}

func FindAndProcessFavicon(domain string) {
	fmt.Printf("   Searching for favicon...\n")

	// Try common favicon paths first
	found := tryCommonFaviconPaths(domain)
	if !found {
		// If not found, parse HTML to find favicon
		found = parseHTMLForFavicon(domain)
	}

	if !found {
		log.Printf("❌ No favicon found for %s\n", domain)
	}
}

func tryCommonFaviconPaths(domain string) bool {
	for _, path := range FaviconPaths {
		faviconURL := normalizeURL(domain) + path
		fmt.Printf("   Trying: %s\n", faviconURL)

		if checkAndProcessImageURL(faviconURL) {
			return true
		}
	}
	return false
}

func parseHTMLForFavicon(domain string) bool {
	data, err := downloader.DownloadFile(domain)
	if err != nil {
		log.Printf("   Error fetching page: %v\n", err)
		return false
	}

	doc, err := goquery.NewDocumentFromReader(strings.NewReader(string(data)))
	if err != nil {
		return false
	}

	// Look for favicon in link tags
	found := false
	doc.Find("link[rel*='icon'], link[rel*='apple-touch-icon']").Each(func(i int, s *goquery.Selection) {
		href, exists := s.Attr("href")
		if exists && href != "" {
			faviconURL := resolveURL(domain, href)
			fmt.Printf("   Found in HTML: %s\n", faviconURL)
			if checkAndProcessImageURL(faviconURL) {
				found = true
			}
		}
	})

	return found
}

func checkAndProcessImageURL(url string) bool {
	if !downloader.CheckURLExists(url) {
		return false
	}

	contentType := downloader.GetContentType(url)
	if strings.HasPrefix(contentType, "image/") {
		processImageURL(url)
		return true
	}

	return false
}

func processImageURL(url string) {
	data, err := downloader.DownloadFile(url)
	if err != nil {
		log.Printf("❌ Error downloading %s: %v\n", url, err)
		return
	}

	if len(data) == 0 {
		log.Printf("❌ Empty response from %s\n", url)
		return
	}

	hasher.CalculateAndDisplayHashes(data, url)
}

func ReadURLsFromFile(filename string) ([]string, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var urls []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line != "" && !strings.HasPrefix(line, "#") {
			urls = append(urls, line)
		}
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return urls, nil
}

func normalizeURL(domain string) string {
	if strings.HasPrefix(domain, "http://") || strings.HasPrefix(domain, "https://") {
		return domain
	}
	return "https://" + domain
}

func resolveURL(base, relative string) string {
	baseURL, err := url.Parse(base)
	if err != nil {
		return relative
	}

	relativeURL, err := url.Parse(relative)
	if err != nil {
		return relative
	}

	return baseURL.ResolveReference(relativeURL).String()
}

File: favmaster/scanner.py
import time
from pathlib import Path
from bs4 import BeautifulSoup
from .downloader import download_file, check_url_exists, get_content_type, normalize_url, resolve_url

# Supported image extensions
SUPPORTED_IMAGE_EXTENSIONS = {
    '.ico', '.png', '.jpg', '.jpeg', '.gif', '.svg', '.webp'
}

# Common favicon paths
FAVICON_PATHS = [
    '/favicon.ico',
    '/favicon.png',
    '/favicon.jpg',
    '/favicon.jpeg',
    '/favicon.svg',
    '/favicon.webp',
    '/images/favicon.ico',
    '/img/favicon.ico',
    '/assets/favicon.ico',
    '/static/favicon.ico',
    '/icon.png',
    '/logo.png',
    '/apple-touch-icon.png',
]

def is_image_url(url):
    """
    Check if URL points to an image based on extension.
    """
    url_lower = url.lower()
    return any(url_lower.endswith(ext) for ext in SUPPORTED_IMAGE_EXTENSIONS)

def find_and_process_favicon(domain):
    """
    Find and process favicon for a domain.
    """
    print("   Searching for favicon...")
    
    # Try common favicon paths first
    found = try_common_favicon_paths(domain)
    if not found:
        # If not found, parse HTML to find favicon
        found = parse_html_for_favicon(domain)
    
    if not found:
        print(f"❌ No favicon found for {domain}")

def try_common_favicon_paths(domain):
    """
    Try common favicon paths for a domain.
    """
    for path in FAVICON_PATHS:
        favicon_url = normalize_url(domain) + path
        print(f"   Trying: {favicon_url}")
        
        if check_and_process_image_url(favicon_url):
            return True
        
        # Small delay to be respectful to servers
        time.sleep(0.1)
    return False

def parse_html_for_favicon(domain):
    """
    Parse HTML to find favicon links.
    """
    try:
        data = download_file(domain)
        soup = BeautifulSoup(data, 'html.parser')
        found = False
        
        # Look for favicon in link tags
        for link in soup.find_all('link', rel=True):
            rel = link.get('rel', [])
            href = link.get('href', '')
            
            # Check for various favicon rel types
            favicon_rels = ['icon', 'shortcut icon', 'apple-touch-icon', 'apple-touch-icon-precomposed']
            if any(icon_type in favicon_rels for icon_type in rel) and href:
                favicon_url = resolve_url(domain, href)
                print(f"   Found in HTML: {favicon_url}")
                if check_and_process_image_url(favicon_url):
                    found = True
        
        return found
    except Exception as e:
        print(f"   Error parsing HTML: {e}")
        return False

def check_and_process_image_url(url):
    """
    Check if URL is an image and process it.
    """
    from .cli import process_single_url
    
    if not check_url_exists(url):
        return False
    
    content_type = get_content_type(url)
    if content_type and content_type.startswith('image/'):
        process_single_url(url)
        return True
    
    return False

def read_urls_from_file(filename):
    """
    Read URLs from a text file.
    """
    urls = []
    try:
        with open(filename, 'r') as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith('#'):
                    urls.append(line)
        return urls
    except Exception as e:
        raise Exception(f"Could not read file {filename}: {e}")
